File: Lecture/Binary Trees/Assignment/isBalanced.java
/*
Time complexity: O(N)
Space complexity: O(H)
where N is the number of nodes in the input tree
and H is the height of the input tree
*/
class Solution {
    class BinaryTreeNode<T> {
        T data;
        BinaryTreeNode<T> left;
        BinaryTreeNode<T> right;
    
        public BinaryTreeNode(T data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public static class CheckBalancedReturnType {
        boolean isBalanced;
        int height;
    }

    public static boolean checkBalanced(BinaryTreeNode<Integer> root) {
        if (root == null) {
            return true;
        }
        return checkBalancedHelper(root).isBalanced;
    }

    public static CheckBalancedReturnType checkBalancedHelper(BinaryTreeNode<Integer> root) {
        if (root == null) {
            CheckBalancedReturnType ans = new CheckBalancedReturnType();
            ans.height = 0;
            ans.isBalanced = true;
            return ans;
        }
        CheckBalancedReturnType leftAns = checkBalancedHelper(root.left);
        CheckBalancedReturnType rightAns = checkBalancedHelper(root.right);
        int heightDiff = leftAns.height > rightAns.height ? (leftAns.height - rightAns.height)
                : (rightAns.height - leftAns.height);
        CheckBalancedReturnType ans;
        if (!leftAns.isBalanced || !rightAns.isBalanced || heightDiff > 1) {
            ans = new CheckBalancedReturnType();
            ans.height = Math.max(leftAns.height, rightAns.height) + 1;
            ans.isBalanced = false;
        } else {
            ans = new CheckBalancedReturnType();
            ans.height = Math.max(leftAns.height, rightAns.height) + 1;
            ans.isBalanced = true;
        }
        return ans;
    }
}

File: Lecture/Binary Trees/Assignment/is-balanced.cpp
/***
 * Given a binary tree, check if it is balanced. Return true if given binary tree is balanced, false otherwise.
 * A empty binary tree or binary tree with zero nodes is always balanced. For a non empty binary tree to be balanced, following conditions must be met:
    1. The left and right subtrees must be balanced.
    2. |hL - hR| <= 1, where hL is the height or depth of left subtree and hR is the height or depth of right subtree.
 ***/
// returns a pair: if tree is balanced or not and it's height
pair<bool, int> isBalancedHelper(BinaryTreeNode<int> *root) {
    pair<bool, int> p;
    if(root == NULL) {
        p.first = true;
        p.second = 0;
        return p;
    }
    
    pair<bool, int> lp = isBalancedHelper(root->left);
    pair<bool, int> rp = isBalancedHelper(root->right);
    
    bool isBalanced = lp.first and rp.first;
    int height = 1 + max(lp.second, rp.second);
    
    if(abs(lp.second - rp.second) > 1)
        isBalanced = false;
    
    p.first = isBalanced;
    p.second = height;
    return p;
}

bool isBalanced(BinaryTreeNode<int> *root) {
	return isBalancedHelper(root).first;
}